File: backend/src/models/PlantingForm.js
const mongoose = require('mongoose');

const plantingFormSchema = new mongoose.Schema({
  farmName: {
    type: String,
    required: true,
    trim: true
  },
  lotNumber: {
    type: String,
    required: true,
    trim: true
  },
  valveNumber: {
    type: String,
    trim: true
  },
  variety: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Variety',
    required: true
  },
  area: {
    type: Number,
    required: true,
    min: 0
  },
  originalArea: {
    type: Number
  },
  plantingDate: {
    type: Date,
    required: true
  },
  expectedHarvestDate: {
    type: Date
  },
  confirmedHarvestDate: {
    type: Date
  },
  status: {
    type: String,
    enum: ['pending', 'approved', 'rejected', 'harvested'],
    default: 'pending'
  },
  approvedBy: {
    type: String,
    trim: true
  },
  approvalDate: {
    type: Date
  },
  approvalPhotos: [{
    type: String
  }],
  observations: {
    type: String,
    trim: true
  },
  producerId: {
    type: String,
    required: true
  },
  notificationSent: {
    type: Boolean,
    default: false
  },
  notificationDate: {
    type: Date
  }
}, {
  timestamps: true
});

// Calculate expected harvest date based on variety cycle
plantingFormSchema.pre('save', async function(next) {
  if (this.isNew || this.isModified('plantingDate') || this.isModified('variety')) {
    const Variety = mongoose.model('Variety');
    const variety = await Variety.findById(this.variety);
    if (variety && this.plantingDate) {
      const expectedDate = new Date(this.plantingDate);
      expectedDate.setDate(expectedDate.getDate() + variety.averageCycleDays);
      this.expectedHarvestDate = expectedDate;
    }
  }
  if (this.isNew && this.area) {
    this.originalArea = this.area;
  }
  next();
});

module.exports = mongoose.model('PlantingForm', plantingFormSchema);

File: mobile/lib/models/planting_form.dart
import 'variety.dart';

class PlantingForm {
  final String id;
  final String farmName;
  final String lotNumber;
  final String? valveNumber;
  final Variety variety;
  final double area;
  final DateTime plantingDate;
  final DateTime? expectedHarvestDate;
  final DateTime? confirmedHarvestDate;
  final String status;
  final String? observations;
  final String producerId;

  PlantingForm({
    required this.id,
    required this.farmName,
    required this.lotNumber,
    this.valveNumber,
    required this.variety,
    required this.area,
    required this.plantingDate,
    this.expectedHarvestDate,
    this.confirmedHarvestDate,
    required this.status,
    this.observations,
    required this.producerId,
  });

  factory PlantingForm.fromJson(Map<String, dynamic> json) {
    return PlantingForm(
      id: json['_id'],
      farmName: json['farmName'],
      lotNumber: json['lotNumber'],
      valveNumber: json['valveNumber'],
      variety: Variety.fromJson(json['variety']),
      area: json['area'].toDouble(),
      plantingDate: DateTime.parse(json['plantingDate']),
      expectedHarvestDate: json['expectedHarvestDate'] != null
          ? DateTime.parse(json['expectedHarvestDate'])
          : null,
      confirmedHarvestDate: json['confirmedHarvestDate'] != null
          ? DateTime.parse(json['confirmedHarvestDate'])
          : null,
      status: json['status'],
      observations: json['observations'],
      producerId: json['producerId'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'farmName': farmName,
      'lotNumber': lotNumber,
      'valveNumber': valveNumber,
      'variety': variety.toJson(),
      'area': area,
      'plantingDate': plantingDate.toIso8601String(),
      'expectedHarvestDate': expectedHarvestDate?.toIso8601String(),
      'confirmedHarvestDate': confirmedHarvestDate?.toIso8601String(),
      'status': status,
      'observations': observations,
      'producerId': producerId,
    };
  }
}
